llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -7482,12 +7482,11 @@ VPRecipeBuilder::tryToWidenMemory(Instruction *I, ArrayRef<VPValue *> Operands,
     VPSingleDefRecipe *VectorPtr;
     if (Reverse) {
       // When folding the tail, we may compute an address that we don't in the
-      // original scalar loop and it may not be inbounds. Drop Inbounds in that
-      // case.
+      // original scalar loop: use none in this case.
       GEPNoWrapFlags Flags =
-          (CM.foldTailByMasking() || !GEP || !GEP->isInBounds())
+          CM.foldTailByMasking() || !GEP
               ? GEPNoWrapFlags::none()
-              : GEPNoWrapFlags::inBounds();
+              : GEP->getNoWrapFlags().withoutNoUnsignedWrap();
```

Comment:
Is this safe to preserve nusw? Doesn't VPVectorEndPointerRecipe add a negative VF offset to the base which can wrap where the scalar one didn't?

Reply:
Oh I see, and we only preserve it when we're not tail folding. So all the computed range would have been covered in the scalar loop.